# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/cairo/cairo/src/cairo-font-options.c
# Commit: 3cdfb39376b7
# Full Hash: 3cdfb39376b736597fe658a322389117686b7e82
# Author: Jonathan Kew <jkew@mozilla.com>
# Date: 2024-05-07 09:13:07
# Regressor Bug: 1892913
# File Overlap Count: 3
# Description:
#   Bug 1892913 - patch 1 - Update cairo source to 1.18.0 r=gfx-reviewers,lsalzman
#   
#   Wholesale replacement of files in gfx/cairo/cairo with contents of the 1.18.0 release tarball,
#   omitting subdirectories that are not part of the core library build.
#   
# ==============================================================================

diff -r 9daa351f875b -r 3cdfb39376b7 gfx/cairo/cairo/src/cairo-font-options.c
--- a/gfx/cairo/cairo/src/cairo-font-options.c	Mon May 06 19:22:35 2024 +0000
+++ b/gfx/cairo/cairo/src/cairo-font-options.c	Mon May 06 19:25:12 2024 +0000
@@ -56,7 +56,10 @@
     CAIRO_HINT_STYLE_DEFAULT,
     CAIRO_HINT_METRICS_DEFAULT,
     CAIRO_ROUND_GLYPH_POS_DEFAULT,
-    NULL
+    NULL, /* variations */
+    CAIRO_COLOR_MODE_DEFAULT,
+    CAIRO_COLOR_PALETTE_DEFAULT,
+    NULL, 0, /* custom palette */
 };
 
 /**
@@ -75,6 +78,10 @@
     options->hint_metrics = CAIRO_HINT_METRICS_DEFAULT;
     options->round_glyph_positions = CAIRO_ROUND_GLYPH_POS_DEFAULT;
     options->variations = NULL;
+    options->color_mode = CAIRO_COLOR_MODE_DEFAULT;
+    options->palette_index = CAIRO_COLOR_PALETTE_DEFAULT;
+    options->custom_palette = NULL;
+    options->custom_palette_size = 0;
 }
 
 void
@@ -88,6 +95,49 @@
     options->hint_metrics = other->hint_metrics;
     options->round_glyph_positions = other->round_glyph_positions;
     options->variations = other->variations ? strdup (other->variations) : NULL;
+    options->color_mode = other->color_mode;
+    options->palette_index = other->palette_index;
+    options->custom_palette_size = other->custom_palette_size;
+    options->custom_palette = NULL;
+    if (other->custom_palette) {
+        options->custom_palette = (cairo_palette_color_t *) malloc (sizeof (cairo_palette_color_t) * options->custom_palette_size);
+        memcpy (options->custom_palette, other->custom_palette, sizeof (cairo_palette_color_t) * options->custom_palette_size);
+    }
+}
+
+cairo_bool_t
+_cairo_font_options_compare (const cairo_font_options_t	*a,
+                             const cairo_font_options_t	*b)
+{
+    if (a->antialias != b->antialias ||
+        a->subpixel_order != b->subpixel_order ||
+        a->lcd_filter != b->lcd_filter ||
+        a->hint_style != b->hint_style ||
+        a->hint_metrics != b->hint_metrics ||
+        a->round_glyph_positions != b->round_glyph_positions ||
+        a->color_mode != b->color_mode ||
+        a->palette_index != b->palette_index ||
+        a->custom_palette_size != b->custom_palette_size)
+    {
+        return FALSE;
+    }
+
+    if (a->variations && b->variations && strcmp (a->variations, b->variations) != 0)
+        return FALSE;
+    else if (a->variations != b->variations)
+        return FALSE;
+
+    if (a->custom_palette && b->custom_palette &&
+        memcmp (a->custom_palette, b->custom_palette, sizeof (cairo_palette_color_t) * a->custom_palette_size) != 0)
+    {
+        return FALSE;
+    }
+    else if (a->custom_palette != b->custom_palette)
+    {
+        return FALSE;
+    }
+
+    return TRUE;
 }
 
 /**
@@ -158,6 +208,7 @@
 _cairo_font_options_fini (cairo_font_options_t *options)
 {
     free (options->variations);
+    free (options->custom_palette);
 }
 
 /**
@@ -186,7 +237,8 @@
  * Checks whether an error has previously occurred for this
  * font options object
  *
- * Return value: %CAIRO_STATUS_SUCCESS or %CAIRO_STATUS_NO_MEMORY
+ * Return value: %CAIRO_STATUS_SUCCESS, %CAIRO_STATUS_NO_MEMORY, or
+ *	%CAIRO_STATUS_NULL_POINTER.
  *
  * Since: 1.0
  **/
@@ -200,7 +252,6 @@
     else
 	return CAIRO_STATUS_SUCCESS;
 }
-slim_hidden_def (cairo_font_options_status);
 
 /**
  * cairo_font_options_merge:
@@ -254,8 +305,18 @@
         options->variations = strdup (other->variations);
       }
     }
+
+    if (other->color_mode != CAIRO_COLOR_MODE_DEFAULT)
+	options->color_mode = other->color_mode;
+    if (other->palette_index != CAIRO_COLOR_PALETTE_DEFAULT)
+	options->palette_index = other->palette_index;
+    if (other->custom_palette) {
+        options->custom_palette_size = other->custom_palette_size;
+        free (options->custom_palette);
+        options->custom_palette = (cairo_palette_color_t *) malloc (sizeof (cairo_palette_color_t) * options->custom_palette_size);
+        memcpy (options->custom_palette, other->custom_palette, sizeof (cairo_palette_color_t) * options->custom_palette_size);
+    }
 }
-slim_hidden_def (cairo_font_options_merge);
 
 /**
  * cairo_font_options_equal:
@@ -290,9 +351,15 @@
 	    options->round_glyph_positions == other->round_glyph_positions &&
             ((options->variations == NULL && other->variations == NULL) ||
              (options->variations != NULL && other->variations != NULL &&
-              strcmp (options->variations, other->variations) == 0)));
+              strcmp (options->variations, other->variations) == 0)) &&
+	    options->color_mode == other->color_mode &&
+	    options->palette_index == other->palette_index &&
+            ((options->custom_palette == NULL && other->custom_palette == NULL) ||
+             (options->custom_palette != NULL && other->custom_palette != NULL &&
+              options->custom_palette_size == other->custom_palette_size &&
+              memcmp (options->custom_palette, other->custom_palette,
+                      sizeof (cairo_palette_color_t) * options->custom_palette_size) == 0)));
 }
-slim_hidden_def (cairo_font_options_equal);
 
 /**
  * cairo_font_options_hash:
@@ -319,13 +386,15 @@
     if (options->variations)
       hash = _cairo_string_hash (options->variations, strlen (options->variations));
 
+    hash ^= options->palette_index;
+
     return ((options->antialias) |
 	    (options->subpixel_order << 4) |
 	    (options->lcd_filter << 8) |
 	    (options->hint_style << 12) |
-	    (options->hint_metrics << 16)) ^ hash;
+	    (options->hint_metrics << 16) |
+            (options->color_mode << 20)) ^ hash;
 }
-slim_hidden_def (cairo_font_options_hash);
 
 /**
  * cairo_font_options_set_antialias:
@@ -346,7 +415,6 @@
 
     options->antialias = antialias;
 }
-slim_hidden_def (cairo_font_options_set_antialias);
 
 /**
  * cairo_font_options_get_antialias:
@@ -389,7 +457,6 @@
 
     options->subpixel_order = subpixel_order;
 }
-slim_hidden_def (cairo_font_options_set_subpixel_order);
 
 /**
  * cairo_font_options_get_subpixel_order:
@@ -412,7 +479,7 @@
 }
 
 /**
- * cairo_font_options_set_lcd_filter:
+ * _cairo_font_options_set_lcd_filter:
  * @options: a #cairo_font_options_t
  * @lcd_filter: the new LCD filter
  *
@@ -422,8 +489,8 @@
  * #cairo_lcd_filter_t for full details.
  **/
 void
-cairo_font_options_set_lcd_filter (cairo_font_options_t *options,
-				   cairo_lcd_filter_t    lcd_filter)
+_cairo_font_options_set_lcd_filter (cairo_font_options_t *options,
+				    cairo_lcd_filter_t    lcd_filter)
 {
     if (cairo_font_options_status (options))
 	return;
@@ -432,7 +499,7 @@
 }
 
 /**
- * cairo_font_options_get_lcd_filter:
+ * _cairo_font_options_get_lcd_filter:
  * @options: a #cairo_font_options_t
  *
  * Gets the LCD filter for the font options object.
@@ -441,7 +508,7 @@
  * Return value: the LCD filter for the font options object
  **/
 cairo_lcd_filter_t
-cairo_font_options_get_lcd_filter (const cairo_font_options_t *options)
+_cairo_font_options_get_lcd_filter (const cairo_font_options_t *options)
 {
     if (cairo_font_options_status ((cairo_font_options_t *) options))
 	return CAIRO_LCD_FILTER_DEFAULT;
@@ -505,7 +572,6 @@
 
     options->hint_style = hint_style;
 }
-slim_hidden_def (cairo_font_options_set_hint_style);
 
 /**
  * cairo_font_options_get_hint_style:
@@ -548,7 +614,6 @@
 
     options->hint_metrics = hint_metrics;
 }
-slim_hidden_def (cairo_font_options_set_hint_metrics);
 
 /**
  * cairo_font_options_get_hint_metrics:
@@ -620,3 +685,187 @@
 {
   return options->variations;
 }
+
+/**
+ * cairo_font_options_set_color_mode:
+ * @options: a #cairo_font_options_t
+ * @color_mode: the new color mode
+ *
+ * Sets the color mode for the font options object. This controls
+ * whether color fonts are to be rendered in color or as outlines.
+ * See the documentation for #cairo_color_mode_t for full details.
+ *
+ * Since: 1.18
+ **/
+cairo_public void
+cairo_font_options_set_color_mode (cairo_font_options_t *options,
+                                   cairo_color_mode_t    color_mode)
+{
+    if (cairo_font_options_status (options))
+	return;
+
+    options->color_mode = color_mode;
+}
+
+/**
+ * cairo_font_options_get_color_mode:
+ * @options: a #cairo_font_options_t
+ *
+ * Gets the color mode for the font options object.
+ * See the documentation for #cairo_color_mode_t for full details.
+ *
+ * Return value: the color mode for the font options object
+ *
+ * Since: 1.18
+ **/
+cairo_public cairo_color_mode_t
+cairo_font_options_get_color_mode (const cairo_font_options_t *options)
+{
+    if (cairo_font_options_status ((cairo_font_options_t *) options))
+	return CAIRO_COLOR_MODE_DEFAULT;
+
+    return options->color_mode;
+}
+
+/**
+ * CAIRO_COLOR_PALETTE_DEFAULT:
+ *
+ * The default color palette index.
+ *
+ * Since: 1.18
+ **/
+
+/**
+ * cairo_font_options_set_color_palette:
+ * @options: a #cairo_font_options_t
+ * @palette_index: the palette index in the CPAL table
+ *
+ * Sets the OpenType font color palette for the font options
+ * object. OpenType color fonts with a CPAL table may contain multiple
+ * palettes. The default color palette index is %CAIRO_COLOR_PALETTE_DEFAULT.
+ *
+ * If @palette_index is invalid, the default palette is used.
+ *
+ * Individual colors within the palette may be overriden with
+ * cairo_font_options_set_custom_palette_color().
+ *
+ * Since: 1.18
+ **/
+void
+cairo_font_options_set_color_palette (cairo_font_options_t *options,
+                                      unsigned int          palette_index)
+{
+    if (cairo_font_options_status (options))
+	return;
+
+    options->palette_index = palette_index;
+}
+
+/**
+ * cairo_font_options_get_color_palette:
+ * @options: a #cairo_font_options_t
+ *
+ * Gets the current OpenType color font palette for the font options object.
+ *
+ * Return value: the palette index
+ *
+ * Since: 1.18
+ **/
+unsigned int
+cairo_font_options_get_color_palette (const cairo_font_options_t *options)
+{
+    if (cairo_font_options_status ((cairo_font_options_t *) options))
+	return CAIRO_COLOR_PALETTE_DEFAULT;
+
+    return options->palette_index;
+}
+
+/**
+ * cairo_font_options_set_custom_palette_color:
+ * @options: a #cairo_font_options_t
+ * @index: the index of the color to set
+ * @red: red component of color
+ * @green: green component of color
+ * @blue: blue component of color
+ * @alpha: alpha component of color
+ *
+ * Sets a custom palette color for the font options object. This
+ * overrides the palette color at the specified color index. This override is
+ * independent of the selected palette index and will remain in place
+ * even if cairo_font_options_set_color_palette() is called to change
+ * the palette index.
+ *
+ * It is only possible to override color indexes already in the font
+ * palette.
+ *
+ * Since: 1.18
+ */
+void
+cairo_font_options_set_custom_palette_color (cairo_font_options_t *options,
+                                             unsigned int index,
+                                             double red, double green,
+                                             double blue, double alpha)
+{
+    unsigned int idx;
+
+    for (idx = 0; idx < options->custom_palette_size; idx++) {
+        if (options->custom_palette[idx].index == index) {
+            break;
+        }
+    }
+
+    if (idx == options->custom_palette_size) {
+        options->custom_palette_size++;
+        options->custom_palette = (cairo_palette_color_t *)
+            _cairo_realloc_ab (options->custom_palette,
+                               sizeof (cairo_palette_color_t),
+                               options->custom_palette_size);
+    }
+
+    /* beware of holes */
+    memset (&options->custom_palette[idx], 0, sizeof (cairo_palette_color_t));
+
+    options->custom_palette[idx].index = index;
+    options->custom_palette[idx].red = red;
+    options->custom_palette[idx].green = green;
+    options->custom_palette[idx].blue = blue;
+    options->custom_palette[idx].alpha = alpha;
+}
+
+/**
+ * cairo_font_options_get_custom_palette_color:
+ * @options: a #cairo_font_options_t
+ * @index: the index of the color to get
+ * @red: return location for red component of color
+ * @green: return location for green component of color
+ * @blue: return location for blue component of color
+ * @alpha: return location for alpha component of color
+ *
+ * Gets the custom palette color for the color index for the font options object.
+ *
+ * Returns: `CAIRO_STATUS_SUCCESS` if a custom palette color is
+ * returned, `CAIRO_STATUS_INVALID_INDEX` if no custom color exists
+ * for the color index.
+ *
+ * Since: 1.18
+ */
+cairo_status_t
+cairo_font_options_get_custom_palette_color (cairo_font_options_t *options,
+                                             unsigned int index,
+                                             double *red, double *green,
+                                             double *blue, double *alpha)
+{
+    unsigned int idx;
+
+    for (idx = 0; idx < options->custom_palette_size; idx++) {
+        if (options->custom_palette[idx].index == index) {
+            *red = options->custom_palette[idx].red;
+            *green = options->custom_palette[idx].green;
+            *blue = options->custom_palette[idx].blue;
+            *alpha = options->custom_palette[idx].alpha;
+            return CAIRO_STATUS_SUCCESS;
+        }
+    }
+
+    return CAIRO_STATUS_INVALID_INDEX;
+}